crates/ide_assists/src/handlers/generate_documentation_template.rs

```diff
@@ -221,18 +223,32 @@ fn self_name(ast_func: &ast::Fn) -> Option<String> {
 }
 
 /// Heper function to get the name of the type of `self`
-fn self_type(ast_func: &ast::Fn) -> Option<String> {
-    ast_func
-        .syntax()
-        .ancestors()
-        .find_map(ast::Impl::cast)
-        .and_then(|i| i.self_ty())
-        .map(|t| (t.to_string()))
+fn self_type(ast_func: &ast::Fn) -> Option<ast::Type> {
+    ast_func.syntax().ancestors().find_map(ast::Impl::cast).and_then(|i| i.self_ty())
```

Comment:
Yes it is deterministic, but actually to be correct it would be:

```rust
ast_func.syntax().parent()?.parent().and_then(ast::Impl::cast).and_then(|i| i.self_ty())
```

Reply:
Oh right, there is an `AssocItemList` in between. Ye in the general case using ancestors here should be fine, don't think there is a configuration where that would give a false positive anyways.

Reply:
I don't think for false positives, but `ancestors` could be slower if not in an `impl`? Well in the AST generally `impl` does not have much ancestors anyway. So I keep `.ancestors.()`?

Reply:
Ye keeping it is fine, source trees are shallow so ancestors won't be that slow